### PR TITLE
Jetty 12 - General cleanup of URIUtil

### PR DESCRIPTION
**Pull request by joakime**
+ Removal of __CHARSET
+ Removal of unused methods
+ Using new switch/case concepts

### CODE REVIEW

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 1]
```diff
@@ -59,9 +58,6 @@
     public static final String HTTP = "http";
     public static final String HTTPS = "https";
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Apparently, only SLASH is used, which is weird anyway (should be inlined everywhere it's used).
The other 2 are not used.
I would remove all 3.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 2]
```diff
@@ -59,9 +58,6 @@
...
 
-    // Use UTF-8 as per http://www.w3.org/TR/html40/appendix/notes.html#non-ascii-chars
-    public static final Charset __CHARSET = StandardCharsets.UTF_8;
-
     /**
      * The characters that are supported by the URI class and that can be decoded by {@link #canonicalPath(String)}
      */
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:59]: Array below should be `private` with upper case name.

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by gregw — APPROVED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 3]
```diff
@@ -206,233 +198,163 @@ private URIUtil()
...
-     * Encode a raw URI String and convert any raw spaces to
-     * their "%20" equivalent.
-     *
-     * @param str input raw string
-     * @return output with spaces converted to "%20"
-     */
-    public static String encodeSpaces(String str)
-    {
-        return StringUtil.replace(str, " ", "%20");
+        return buf == null ? path : buf.toString();
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Wow the logic of this method is really hard to follow, and I think unnecessary.

First, we loop once over the path; if we find a reserved character, we exit immediately, to enter the second loop where we restart from index 0; in the second loop, if we find a CTL or non-ASCII character we convert the path to UTF-8 bytes for the second time, exit and enter a 3rd loop where we restart from index 0.

I understand we want to optimize the common path with no allocations (so just one loop over all the chars).
But the less common case should be written better, iterating only once.

The new algorithm should be something like:
iterate over the chars; if a special is found, allocate a `StringBuilder`, copy all the non-reserved chars so far, and then append the % representation of the reserved char. If we find a CTL or non-ASCII, we keep what we have so far, convert to UTF-8 bytes the rest, and continue the loop until the end.

**Review by sbordet — CHANGES_REQUESTED** [review]
Also all the javadocs `<code>` should be converted to `{@code}`

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 3 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: I simplified encodePath() a great deal.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 4]
```diff
@@ -197,242 +189,92 @@
...
+     * the '{@code /}' character is not encoded.
      *
-     * @param path The path the encode
+     * @param path The path to encode
      * @return The encoded path
      */
     public static String encodePath(String path)
     {
         if (path == null || path.length() == 0)
             return path;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:221]: Use `StringUtil.isEmpty(path)`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 5]
```diff
@@ -197,242 +189,92 @@
...
+    static
+    {
+        ENCODE_PATH_NEEDS_ENCODING = new boolean[128];
+        // Special characters
+        for (char c: "%?;#\"'<> [\\]^`{|}".toCharArray())
+            ENCODE_PATH_NEEDS_ENCODING[c] = true;
+        // control characters
+        ENCODE_PATH_NEEDS_ENCODING[0x7f] = true;
+        for (int i = 0; i < 0x20; i++)
+            ENCODE_PATH_NEEDS_ENCODING[i] = true;
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:207]: I believe SPACE 0x20 should be encoded too.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 6]
```diff
@@ -197,242 +189,92 @@
...
-                            bytes = path.getBytes(URIUtil.__CHARSET);
-                            buf = new StringBuilder(path.length() * 2);
-                            break loop;
-                        }
-                }
-            }
-            if (buf == null)
-                return null;
+            char c = path.charAt(i);
+            if (c > 0x7F) // 8-bit +
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:231]: 0x7F should be encoded too, so `c >= 0x7F` here.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 7]
```diff
@@ -197,242 +189,92 @@
...
+                needsEncoding = true;
         }
 
-        int i;
+        if (needsByteEncoding)
+            return encodePathBytes(path);
+        else if (needsEncoding)
+        {
+            return encodePathString(path);
+        }
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Drop the braces.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java [hunk 8]
```diff
@@ -518,6 +518,21 @@ public static boolean isHex(String str, int offset, int len)
...
+            buf.append((char)((d > 9 ? ('A' - 10) : '0') + d));
+            d = 0xf & c;
+            buf.append((char)((d > 9 ? ('A' - 10) : '0') + d));
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java]: No need for this method (see comment in URIUtil).

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 9]
```diff
@@ -197,242 +189,92 @@
...
-
-                default:
-                    if (c < 0x20 || c >= 0x7f)
-                    {
-                        bytes = path.getBytes(URIUtil.__CHARSET);
-                        break loop;
-                    }
-                    buf.append(c);
+                buf.append('%');
+                TypeUtil.toHex(c, buf);
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: You can safely cast `c` to a `byte` here, since this method is called only if all chars are ASCII, therefore bytes.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 10]
```diff
@@ -197,242 +189,92 @@
...
-                            TypeUtil.toHex(c, buf);
-                        }
-                        else
-                            buf.append((char)c);
-                }
+                buf.append('%');
+                TypeUtil.toHex(b, buf);
             }
+            else
+                buf.append((char)b);
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:284]: Add braces.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 11]
```diff
@@ -1582,32 +1428,29 @@ public static void appendSchemeHostPort(StringBuilder url, String scheme, String
...
+                    break;
+
+                default:
+                    url.append(':').append(port);
             }
         }
     }
 
-    // Only URIUtil is using this method
+    // Only used by URIUtil
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Make it private then.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 12]
```diff
@@ -1622,6 +1465,7 @@ static boolean equalsIgnoreEncodings(String uriA, String uriB)
         }
     }
 
+    // Only used by URIUtil
```
Comment (sbordet) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Make it private then.

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java [hunk 13]
```diff
@@ -1045,7 +1021,7 @@ public void testEncodeDecodeVisibleOnly()
         {
             assertTrue(c > 0x20 && c < 0x80);
```
Comment (sbordet) [inline jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java]: Should not `c < 0x7F`, since `7F` is encoded and can never be present as a character.
Also, the comment 2 lines above has a typo: `endoded`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 5 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:207]: Already covered in the special chars above

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 6 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:231]: already covered in ENCODE_PATH_NEEDS_ENCODING

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 11 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Can't make it private, as it's tested in URIUtilTest.
And protected isn't valid.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 12 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Can't make it private, as it's tested in URIUtilTest.
And protected isn't valid.

File: jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java (hunk 13 above)
Comment (joakime) [reply jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java]: fair enough, but the check for control characters is checked 2 lines below this, which would have tripped on the raw 0x7f usage as well.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 6 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:231]: added testcase specific for this to show it's protected

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 5 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:207]: added testcase specific for this to show it's protected

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 11 above)
Comment (sbordet) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Then the comment is wrong, should be `Only used by URIUtil*Test*.`

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 12 above)
Comment (sbordet) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Then the comment is wrong, should be `Only used by URIUtil*Test*.`

Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: It's used by URIUtil as a sub-method, and that sub-method is tested in URIUtilTest, it's not exposed to others.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 6 above)
Comment (sbordet) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:231]: Doh, you're right.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 12 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Went ahead and cleaned it out

**Review by sbordet — APPROVED** [review]
